### PR TITLE
Fix/template overloads (recreation of #255)

### PR DESCRIPTION
**Pull request by DerThorsten**
recreation of #255 as branch instead of fork

### CODE REVIEW

File: cpp/arcticdb/processing/operation_types.hpp [hunk 1]
```diff
@@ -389,10 +389,11 @@ bool operator()(int64_t t, const std::unordered_set<uint64_t>& u) const {
     else
         return u.count(t) > 0;
 }
-// This is the version called when checking string set membership
```
Comment (qc00) [inline cpp/arcticdb/processing/operation_types.hpp:392]: Could you share what's calling this?

Comment (DerThorsten) [reply cpp/arcticdb/processing/operation_types.hpp:392]:
```
In file included from /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.cpp:9:
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.hpp:83:32: error: no matching function for call to object of type 'arcticdb::IsInOperator'
                              if(func(offset, offset_set))
                                 ^~~~
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.hpp:142:45: note: in instantiation of function template specialization 'arcticdb::binary_membership<arcticdb::IsInOperator &&>' requested here
              return transform_to_placeholder(binary_membership<decltype(func)>(l, *r, std::forward<Func>(func)));
                                              ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.cpp:150:20: note: in instantiation of function template specialization 'arcticdb::visit_binary_membership<arcticdb::IsInOperator>' requested here
              return visit_binary_membership(left, right, IsInOperator{});
                     ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_types.hpp:380:6: note: candidate function not viable: no known conversion from 'emilib::HashSet<position_t>' (aka 'HashSet<long>') to 'const std::unordered_set<int64_t>' (aka 'const unordered_set<long long>') for 2nd argument
  bool operator()(uint64_t t, const std::unordered_set<int64_t>& u) const {
       ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_types.hpp:386:6: note: candidate function not viable: no known conversion from 'emilib::HashSet<position_t>' (aka 'HashSet<long>') to 'const std::unordered_set<uint64_t>' (aka 'const unordered_set<unsigned long long>') for 2nd argument
  bool operator()(int64_t t, const std::unordered_set<uint64_t>& u) const {
       ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_types.hpp:392:6: note: candidate function not viable: no known conversion from 'HashSet<arcticdb::entity::position_t>' to 'const HashSet<int64_t>' for 2nd argument
  bool operator()(int64_t t, const emilib::HashSet<int64_t>& u) const {
       ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_types.hpp:377:6: note: candidate template ignored: could not match 'unordered_set' against 'HashSet'
  bool operator()(T t, const std::unordered_set<U>& u) const {
       ^
  In file included from /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.cpp:9:
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.hpp:60:127: warning: lambda capture 'column_desc_tag' is not used [-Wunused-lambda-capture]
              entity::details::visit_type(value_set.base_type().data_type(),[&column_with_strings, &value_set, &func, &output, &column_desc_tag] (auto value_set_desc_tag) {
                                                                                                                                ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/entity/types-inl.hpp:51:9: note: in instantiation of function template specialization 'arcticdb::binary_membership(const arcticdb::ColumnWithStrings &, arcticdb::ValueSet &, arcticdb::IsNotInOperator &&)::(anonymous class)::operator()<arcticdb::entity::DataTypeTag<DataType::UINT8>>' requested here
          DT_CASE(UINT8)
          ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/entity/types-inl.hpp:50:12: note: expanded from macro 'DT_CASE'
      return c(DataTypeTag<DataType::__T__>());
             ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.hpp:56:26: note: in instantiation of function template specialization 'arcticdb::entity::details::visit_type<(lambda at /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.hpp:56:85)>' requested here
          entity::details::visit_type(column_with_strings.column_->type().data_type(),[&column_with_strings, &value_set, &func, &output] (auto column_desc_tag) {
                           ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.hpp:142:45: note: in instantiation of function template specialization 'arcticdb::binary_membership<arcticdb::IsNotInOperator &&>' requested here
              return transform_to_placeholder(binary_membership<decltype(func)>(l, *r, std::forward<Func>(func)));
                                              ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.cpp:152:20: note: in instantiation of function template specialization 'arcticdb::visit_binary_membership<arcticdb::IsNotInOperator>' requested here
              return visit_binary_membership(left, right, IsNotInOperator{});
                     ^
  In file included from /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.cpp:9:
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.hpp:60:127: warning: lambda capture 'column_desc_tag' is not used [-Wunused-lambda-capture]
              entity::details::visit_type(value_set.base_type().data_type(),[&column_with_strings, &value_set, &func, &output, &column_desc_tag] (auto value_set_desc_tag) {
                                                                                                                                ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/entity/types-inl.hpp:52:9: note: in instantiation of function template specialization 'arcticdb::binary_membership(const arcticdb::ColumnWithStrings &, arcticdb::ValueSet &, arcticdb::IsNotInOperator &&)::(anonymous class)::operator()<arcticdb::entity::DataTypeTag<DataType::UINT16>>' requested here
          DT_CASE(UINT16)
          ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/entity/types-inl.hpp:50:12: note: expanded from macro 'DT_CASE'
      return c(DataTypeTag<DataType::__T__>());
             ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.hpp:56:26: note: in instantiation of function template specialization 'arcticdb::entity::details::visit_type<(lambda at /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.hpp:56:85)>' requested here
          entity::details::visit_type(column_with_strings.column_->type().data_type(),[&column_with_strings, &value_set, &func, &output] (auto column_desc_tag) {
                           ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.hpp:142:45: note: in instantiation of function template specialization 'arcticdb::binary_membership<arcticdb::IsNotInOperator &&>' requested here
              return transform_to_placeholder(binary_membership<decltype(func)>(l, *r, std::forward<Func>(func)));
                                              ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.cpp:152:20: note: in instantiation of function template specialization 'arcticdb::visit_binary_membership<arcticdb::IsNotInOperator>' requested here
              return visit_binary_membership(left, right, IsNotInOperator{});
                     ^
  In file included from /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.cpp:9:
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.hpp:60:127: warning: lambda capture 'column_desc_tag' is not used [-Wunused-lambda-capture]
              entity::details::visit_type(value_set.base_type().data_type(),[&column_with_strings, &value_set, &func, &output, &column_desc_tag] (auto value_set_desc_tag) {
                                                                                                                                ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/entity/types-inl.hpp:53:9: note: in instantiation of function template specialization 'arcticdb::binary_membership(const arcticdb::ColumnWithStrings &, arcticdb::ValueSet &, arcticdb::IsNotInOperator &&)::(anonymous class)::operator()<arcticdb::entity::DataTypeTag<DataType::UINT32>>' requested here
          DT_CASE(UINT32)
          ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/entity/types-inl.hpp:50:12: note: expanded from macro 'DT_CASE'
      return c(DataTypeTag<DataType::__T__>());
             ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.hpp:56:26: note: in instantiation of function template specialization 'arcticdb::entity::details::visit_type<(lambda at /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.hpp:56:85)>' requested here
          entity::details::visit_type(column_with_strings.column_->type().data_type(),[&column_with_strings, &value_set, &func, &output] (auto column_desc_tag) {
                           ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.hpp:142:45: note: in instantiation of function template specialization 'arcticdb::binary_membership<arcticdb::IsNotInOperator &&>' requested here
              return transform_to_placeholder(binary_membership<decltype(func)>(l, *r, std::forward<Func>(func)));
                                              ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.cpp:152:20: note: in instantiation of function template specialization 'arcticdb::visit_binary_membership<arcticdb::IsNotInOperator>' requested here
              return visit_binary_membership(left, right, IsNotInOperator{});
                     ^
  In file included from /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.cpp:9:
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.hpp:60:127: warning: lambda capture 'column_desc_tag' is not used [-Wunused-lambda-capture]
              entity::details::visit_type(value_set.base_type().data_type(),[&column_with_strings, &value_set, &func, &output, &column_desc_tag] (auto value_set_desc_tag) {
                                                                                                                                ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/entity/types-inl.hpp:54:9: note: in instantiation of function template specialization 'arcticdb::binary_membership(const arcticdb::ColumnWithStrings &, arcticdb::ValueSet &, arcticdb::IsNotInOperator &&)::(anonymous class)::operator()<arcticdb::entity::DataTypeTag<DataType::UINT64>>' requested here
          DT_CASE(UINT64)
          ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/entity/types-inl.hpp:50:12: note: expanded from macro 'DT_CASE'
      return c(DataTypeTag<DataType::__T__>());
             ^
  /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.hpp:56:26: note: in instantiation of function template specialization 'arcticdb::entity::details::visit_type<(lambda at /Users/thorstenbeier/src/ArcticDB/cpp/arcticdb/processing/operation_dispatch_binary.hpp:56:85)>' requested here
          entity::details::visit_type(column_with_strings.column_->type().data_type(),[&column_with_strings, &value_set, &func, &output] (auto column_desc_tag) {

.... [EVEN MORE]
```

Comment (qc00) [reply cpp/arcticdb/processing/operation_types.hpp:392]: I am very sceptical of the fact that the compiler is not willing to do the type coercion, but the `emilib contains()` accepted it.

We're replacing `emilib` anyway, so maybe this mystery don't have to be solved.

Comment (DerThorsten) [reply cpp/arcticdb/processing/operation_types.hpp:392]: can we anyhow merge/approve this? or is `emilib` removed very soon?

**Review by qc00 — APPROVED** [review]

File: cpp/arcticdb/processing/operation_types.hpp (hunk 1 above)
Comment (qc00) [reply cpp/arcticdb/processing/operation_types.hpp:392]: William is working on it, but I'll unblock you for now

Comment (DerThorsten) [reply cpp/arcticdb/processing/operation_types.hpp:392]: thanks!

**Pull request merged**